mautic/mautic
# Form enhancements - submit actions for campaign forms, new re-post submit action, and improved support for developers and more

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y
| Related user documentation PR URL | todo
| Related developer documentation PR URL | todo
| Issues addressed (#s or URLs) | #1121 #317 #900 #761 #1790 #2249
| BC breaks? | n
| Deprecations? | y

This PR includes the following PRs that need to be merged prior to this:

- [x] #2422 - bug fix 
- [x] #2427 - included to work out conflicts with the other PR
- [ ] #2558 - this is not part of this PR but a change from it is needed to resolve conflicts; once it's merged, I'll update this one to resolve the conflict and use the new trait that's part of that PR
- [x] #2351 - merged into this PR and closed as it crossed over with many of the changes made by this PR. 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds the following enhancements:

1. It brings back support for submit actions to campaign forms. The developer now has the option to configure if a submit action should be supported by a campaign form or not. This allows us to only make submit actions available to campaign forms that make sense - i.e. asset downloads, send form result emails, etc.

2. It improves form submit actions, fields, and adds a new validator feature for developers. Instead of using a static `callback`, the same method is now used as campaign events - configure an `eventName` that gets dispatched. `callback` BC was left in place as this PR does not (yet) refactor all core form actions to this new method. It also added a new option to register a validator that can be used to validate all or specific types of form fields.

3. This PR adds a new feature to re-post form results to a 3rd party form/script. This is the main reason I brought back submit actions for campaign forms. The new submit action allows configuration of a post URL, authentication header (if applicable), a fallback email if the form fails to re-post to the configured URL for whatever reason, and the option to map Mautic's form fields to a different name for the receiving script. 

4. It adds support for honey pot captcha. Just leave the Captcha Answer blank to treat the captcha as a honeypot.

5. Form page breaks! Use the new "Page break" form field to insert pages into the form that injects previous (optional) or continue buttons to progress through a form. Mautic's JS automatically handles advancing form pages.

6. Now supports contact and form field tokens in the post submit redirect/message.

7. Adds support for defining label and value for select/radio groups - this has also been applied to contact fields.

8. Allows basic formatting HTML for form field labels (including radio/checkbox labels)

9. Adds new option to sync the form's select/checkbox/radio list with a contact field.

#### Steps to test this PR:
1. For the campaign submit actions - edit or create a campaign form. Click the Actions tab and select the dropdown for actions. Notice that only a handful are available (download asset, etc). Add a new action then submit the form - the action should be executed. Now clone a standalone form with that has the "Modify Contact's Segments" and select the clone to be a campaign form. Notice that the action was removed. Also notice that the available action list is limited. Upon saving - it should not be saved as part of the new form. Now do the opposite and clone a campaign form, select the new clone to be a standalone, then view the action drop down list - it should be the full list. Editing a campaign or standalone form should result in the appropriate actions in the dropdown list. 
 
2. Ensure that form actions still work. `Send form results` and the new `post results to a new form` use the new method - the others use the old. So test a submit action with those two plus downloads asset.

3. Test the new post to another form submit action by posting to a http://requestb.in/. The results should show after a form submit. Use a bogus URL to test the failed email feature.

4. Inject two captchas. One with an answer and one without. When previewing the form, manually manipulate the value of the one without an answer. Submit the form and view the DOM through the dev console. Notice that the captcha with an answer (that was left blank) failed and also the one without an answer (with a value manually injected) failed. Now refresh the page and enter the answer for the visible captcha and leave the honeypot one alone. The form should advance.

5. Create a form with several form fields. Add a new page break (or more) and order them where a page break is desired. Save and view the form - there show now be a previous button (if enabled in by entering a label on the properties tab of the form) and a continue button. Advance through the form to the end. Notice that last continue button was replaced by the submit button. The form should continue to validate, advance back and forth, and of course submit.

6. Use a {contactfield=x} or {formfield=x} token in the post submit redirect. It should be properly replaced on a form submit.

7. Important one - ensure form field lists and contact field lists continue to work. Existing fields should continue to work. Reordering options in lists should still work. 

8. Use basic HTML (b, a, etc) in a form field's label and/or list labels and save - they should now display. Try using a `<script>alert('hi');</script` as a label and it should not inject.

9. Create a custom select contact field with some list options. Create a new form select field and map the created contact field. Set the list to sync on the properties tab and save. The list should be populated with that of the contacts. If you edit the contact custom field and rebuild the form  - the select options should be updated.

#### List deprecations along with the new alternative:
1. This PR deprecates some keys for registering submit actions with the FormBuilderEvent. Instead of using a `callback`, one should now use `eventName` and create a listener for that event `$event->addSubmitAction(['eventName' => AcmeEvents::SUBMIT_ACTION, ...])`. 
2. The submit action's `validator` callback has been deprecated - it was out of place. Instead use the new `FormBuilderEvent::addValidator(['eventName' => AcmeEvents::VALIDATOR])`. Add a listener for `AcmeEvents::THE_EVENT` that handles the validation. 
3. With the old `callback` reflection method, a method could return an array with another `callback` defined that was executed after all submit actions were processed (like Assets to inject a response to initiate a download). With the move toward event based submit actions, when the action is processed by the listener for `AcmeEvents::SUBMIT_ACTION`, either inject the response at that time with `SubmissionEvent::setPostSubmitCallbackResponse()` or register a callback to be executed after all submit actions have been processed by using `SubmissionEvent::setPostSubmitCallback($key, ['eventName' => AcmeEvents::CALLBACK])`
4. This also includes some cleanups for registering CampaignEvents, so `CampaignBuilderEvent::addLeadDecision`, `getLeadDecisions`, `addLeadCondition` and `getLeadConditions` have been deprecated. Use the same method without `Lead` i.e. `addDecision`
